src/main/java/picard/sam/SplitSamBySize.java

```diff
@@ -0,0 +1,152 @@
+ */
+@CommandLineProgramProperties(
+        summary = SplitSamBySize.USAGE_SUMMARY + SplitSamBySize.USAGE_DETAILS,
+        oneLineSummary = SplitSamBySize.USAGE_SUMMARY,
+        programGroup = SamOrBam.class)
+@DocumentedFeature
+public class SplitSamBySize extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input query-grouped SAM/BAM file into multiple BAM files " +
+            "while maintaining the sort order. This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SplitSamBySize \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000 \\ <br />" +
+            "     SPLIT_TO_N_READS=48000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N_READS", doc = "Split to have approximately N reads per output file.", optional = true,
+            mutex = {"SPLIT_TO_N_FILES"})
+    public int SPLIT_TO_N_READS;
+
+    @Argument(shortName = "N_FILES", doc = "Split to N files.", optional = true, mutex = {"SPLIT_TO_N_READS"})
+    public int SPLIT_TO_N_FILES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = "O", doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT_DIR;
+
+    @Argument(shortName = "OUT_PREFIX", doc = "Prefix to name output files.")
+    public String OUT_PREFIX = "shard";
+
+    private final Log log = Log.getInstance(SplitSamBySize.class);
+
+    public static void main(final String[] argv) {
+        System.exit(new SplitSamBySize().instanceMain(argv));
+    }
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+        final SAMFileHeader header = reader.getFileHeader();
+        if (header.getSortOrder() == SAMFileHeader.SortOrder.coordinate) {
+            log.warn("Splitting a coordinate sorted bam may result in invalid bams " +
+                    "that do not always contain each read's mate in the same bam.");
+        }
+
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+
+        int splitToNFiles = SPLIT_TO_N_FILES != 0 ? SPLIT_TO_N_FILES : (int) Math.ceil(TOTAL_READS_IN_INPUT / SPLIT_TO_N_READS);
+        int readsPerFile = (int) Math.ceil(TOTAL_READS_IN_INPUT / (double) splitToNFiles);
+        SAMFileWriter[] writers = generateWriters(factory, header, splitToNFiles);
```

Comment:
What's the advantage of setting up the writers in advance? As far as I can see the only actual use of knowing how many writers you'll have is that you are able to put the "correct" amount of zeros in the output filename....but even that doesn't require that you create the writers in advance. If on the other hand you generate the writers on the fly, the tool would be able to by used without knowledge of the size of the input...